
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://tschuelia.github.io/PyPythia/latest/api/prediction/">
      
      
        <link rel="prev" href="../raxmlng/">
      
      
        <link rel="next" href="../predictor/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.7">
    
    
      
        <title>prediction - PyPythia</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#pypythia.prediction.collect_features" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="PyPythia" class="md-header__button md-logo" aria-label="PyPythia" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PyPythia
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              prediction
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9 2c3.87 0 7 3.13 7 7 0 2.38-1.19 4.47-3 5.74V17c0 .55-.45 1-1 1H6c-.55 0-1-.45-1-1v-2.26C3.19 13.47 2 11.38 2 9c0-3.87 3.13-7 7-7M6 21v-1h6v1c0 .55-.45 1-1 1H7c-.55 0-1-.45-1-1M9 4C6.24 4 4 6.24 4 9c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58 0-2.76-2.24-5-5-5m10 9h-2l-3.2 9h1.9l.7-2h3.2l.7 2h1.9zm-2.15 5.65L18 15l1.15 3.65z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/tschuelia/PyPythia" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tschuelia/PyPythia
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="PyPythia" class="md-nav__button md-logo" aria-label="PyPythia" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    PyPythia
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/tschuelia/PyPythia" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    tschuelia/PyPythia
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    User Guide
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../msa/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    msa
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../raxmlng/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    raxmlng
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    prediction
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    prediction
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pypythia.prediction.collect_features" class="md-nav__link">
    <span class="md-ellipsis">
      collect_features
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pypythia.prediction.predict_difficulty" class="md-nav__link">
    <span class="md-ellipsis">
      predict_difficulty
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../predictor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    predictor
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../custom_types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    custom_types
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../custom_errors/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    custom_errors
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    config
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pypythia.prediction.collect_features" class="md-nav__link">
    <span class="md-ellipsis">
      collect_features
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pypythia.prediction.predict_difficulty" class="md-nav__link">
    <span class="md-ellipsis">
      predict_difficulty
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>prediction</h1>

<div class="doc doc-object doc-function">


<h2 id="pypythia.prediction.collect_features" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pypythia</span><span class="o">.</span><span class="n">prediction</span><span class="o">.</span><span class="n">collect_features</span><span class="p">(</span><span class="n">msa</span><span class="p">,</span> <span class="n">msa_file</span><span class="p">,</span> <span class="n">raxmlng</span><span class="p">,</span> <span class="n">pars_trees_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log_info</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Helper function to collect all features required for predicting the difficulty of the MSA.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>msa</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.msa.MSA" href="../msa/#pypythia.msa.MSA">MSA</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>MSA object corresponding to the MSA file to compute the features for.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>raxmlng</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.raxmlng.RAxMLNG" href="../raxmlng/#pypythia.raxmlng.RAxMLNG">RAxMLNG</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Initialized RAxMLNG object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pars_trees_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to store the inferred parsimony trees. Defaults to None.
In this case, the trees are not stored.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>log_info</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, log intermediate progress information using the default logger. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threads</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of threads to use for parallel parsimony tree inference. Defaults to None.
Uses the RAxML-NG auto parallelization scheme if none is set.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed to use for the parsimony tree inference. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    Dataframe containing a single row with all features required for predicting the difficulty of the MSA.
    The columns correspond to the feature names the predictor was trained with.</p>

            <details class="quote">
              <summary>Source code in <code>pypythia/prediction.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">collect_features</span><span class="p">(</span>
    <span class="n">msa</span><span class="p">:</span> <span class="n">MSA</span><span class="p">,</span>
    <span class="n">msa_file</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">,</span>
    <span class="n">raxmlng</span><span class="p">:</span> <span class="n">RAxMLNG</span><span class="p">,</span>
    <span class="n">pars_trees_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">log_info</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Helper function to collect all features required for predicting the difficulty of the MSA.</span>

<span class="sd">    Args:</span>
<span class="sd">        msa (MSA): MSA object corresponding to the MSA file to compute the features for.</span>
<span class="sd">        raxmlng (RAxMLNG): Initialized RAxMLNG object.</span>
<span class="sd">        pars_trees_file (pathlib.Path, optional): Path to store the inferred parsimony trees. Defaults to None.</span>
<span class="sd">            In this case, the trees are not stored.</span>
<span class="sd">        log_info (bool, optional): If True, log intermediate progress information using the default logger. Defaults to False.</span>
<span class="sd">        threads (int, optional): The number of threads to use for parallel parsimony tree inference. Defaults to None.</span>
<span class="sd">            Uses the RAxML-NG auto parallelization scheme if none is set.</span>
<span class="sd">        seed (int, optional): Random seed to use for the parsimony tree inference. Defaults to 0.</span>
<span class="sd">    Returns:</span>
<span class="sd">        Dataframe containing a single row with all features required for predicting the difficulty of the MSA.</span>
<span class="sd">        The columns correspond to the feature names the predictor was trained with.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmpdir</span><span class="p">:</span>
        <span class="n">msa_file</span> <span class="o">=</span> <span class="n">msa_file</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">msa</span><span class="o">.</span><span class="n">get_raxmlng_model</span><span class="p">()</span>

        <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span><span class="s2">&quot;Retrieving num_taxa, num_sites.&quot;</span><span class="p">)</span>

        <span class="n">n_pars_trees</span> <span class="o">=</span> <span class="mi">24</span>
        <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Inferring </span><span class="si">{</span><span class="n">n_pars_trees</span><span class="si">}</span><span class="s2"> parsimony trees with random seed </span><span class="si">{</span><span class="n">seed</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">trees</span> <span class="o">=</span> <span class="n">raxmlng</span><span class="o">.</span><span class="n">infer_parsimony_trees</span><span class="p">(</span>
            <span class="n">msa_file</span><span class="p">,</span>
            <span class="n">model</span><span class="p">,</span>
            <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">tmpdir</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;pars&quot;</span><span class="p">,</span>
            <span class="n">redo</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
            <span class="n">n_trees</span><span class="o">=</span><span class="n">n_pars_trees</span><span class="p">,</span>
            <span class="o">**</span><span class="nb">dict</span><span class="p">(</span><span class="n">threads</span><span class="o">=</span><span class="n">threads</span><span class="p">)</span> <span class="k">if</span> <span class="n">threads</span> <span class="k">else</span> <span class="p">{},</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">pars_trees_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Storing the inferred parsimony trees in the file </span><span class="si">{</span><span class="n">pars_trees_file</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">trees</span><span class="p">,</span> <span class="n">pars_trees_file</span><span class="p">)</span>

        <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
            <span class="s2">&quot;Computing the RF-Distance for the parsimony trees.&quot;</span>
        <span class="p">)</span>
        <span class="n">num_topos</span><span class="p">,</span> <span class="n">rel_rfdist</span> <span class="o">=</span> <span class="n">raxmlng</span><span class="o">.</span><span class="n">get_rfdistance_results</span><span class="p">(</span><span class="n">trees</span><span class="p">,</span> <span class="n">redo</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

        <span class="n">features</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;num_taxa&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_taxa</span><span class="p">,</span>
            <span class="s2">&quot;num_sites&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_sites</span><span class="p">,</span>
            <span class="s2">&quot;num_patterns&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_patterns</span><span class="p">,</span>
            <span class="s2">&quot;num_patterns/num_taxa&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_patterns</span> <span class="o">/</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_taxa</span><span class="p">,</span>
            <span class="s2">&quot;num_sites/num_taxa&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_sites</span> <span class="o">/</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_taxa</span><span class="p">,</span>
            <span class="s2">&quot;num_patterns/num_sites&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_patterns</span> <span class="o">/</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_sites</span><span class="p">,</span>
            <span class="s2">&quot;proportion_gaps&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">proportion_gaps</span><span class="p">,</span>
            <span class="s2">&quot;proportion_invariant&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">proportion_invariant</span><span class="p">,</span>
            <span class="s2">&quot;entropy&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">entropy</span><span class="p">(),</span>
            <span class="s2">&quot;bollback&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">bollback_multinomial</span><span class="p">(),</span>
            <span class="s2">&quot;pattern_entropy&quot;</span><span class="p">:</span> <span class="n">msa</span><span class="o">.</span><span class="n">pattern_entropy</span><span class="p">(),</span>
            <span class="s2">&quot;avg_rfdist_parsimony&quot;</span><span class="p">:</span> <span class="n">rel_rfdist</span><span class="p">,</span>
            <span class="s2">&quot;proportion_unique_topos_parsimony&quot;</span><span class="p">:</span> <span class="n">num_topos</span> <span class="o">/</span> <span class="n">n_pars_trees</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pypythia.prediction.predict_difficulty" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pypythia</span><span class="o">.</span><span class="n">prediction</span><span class="o">.</span><span class="n">predict_difficulty</span><span class="p">(</span><span class="n">msa_file</span><span class="p">,</span> <span class="n">model_file</span><span class="o">=</span><span class="n">DEFAULT_MODEL_FILE</span><span class="p">,</span> <span class="n">raxmlng</span><span class="o">=</span><span class="n">DEFAULT_RAXMLNG_EXE</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">file_format</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">deduplicate</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">remove_full_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">result_prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">store_results</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log_info</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents first">

        <p>Predict the difficulty of an MSA using the PyPythia difficulty predictor.</p>
<p>Per default, the MSA is deduplicated and full gap sequences are removed before the difficulty is predicted.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>msa_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the MSA file. Note that the MSA file must be in either FASTA or PHYLIP format.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>model_file</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the trained difficulty predictor model.
Defaults to the latest model shipped with PyPythia.</p>
              </div>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.config.DEFAULT_MODEL_FILE" href="../config/#pypythia.config.DEFAULT_MODEL_FILE">DEFAULT_MODEL_FILE</a></code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>raxmlng</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the RAxML-NG executable.
If not set, uses the RAxML-NG binary found in the PATH environment variable.</p>
              </div>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.config.DEFAULT_RAXMLNG_EXE" href="../config/#pypythia.config.DEFAULT_RAXMLNG_EXE">DEFAULT_RAXMLNG_EXE</a></code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threads</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of threads to use for parallel parsimony tree inference. If not set, uses the
RAxML-NG auto parallelization scheme.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Random seed to use for the parsimony tree inference. Defaults to 0.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_format</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.custom_types.FileFormat" href="../custom_types/#pypythia.custom_types.FileFormat">FileFormat</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>File format of the MSA file. Defaults to None. In this case, the file format
is inferred based on the file content. See <code>pypythia.msa.parse_msa</code> for information on when this is required.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>data_type</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="pypythia.custom_types.DataType" href="../custom_types/#pypythia.custom_types.DataType">DataType</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data type of the MSA sequences. Defaults to None. In this case, the data type
is inferred based on the file content. See <code>pypythia.msa.parse_msa</code> for information on when this is required.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>deduplicate</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, remove duplicate sequences from the MSA. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>remove_full_gaps</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, remove full gap sequences from the MSA. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>result_prefix</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Prefix for the result files. Defaults to None. In this case, the prefix
is set to the MSA file name.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>store_results</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, store intermediate results as file. Defaults to True.
In this case, the following files are stored:
- The reduced MSA in PHYLIP format (if duplicates or full gap sequences were removed) in <code>{result_prefix}.reduced.phy</code>
- The inferred parsimony trees in Newick format in <code>{result_prefix}.pythia.trees</code>
- The shapley values as waterfall plot in <code>{result_prefix}.shap.pdf</code>
- The features and predicted difficulty as CSV file in <code>{result_prefix}.pythia.csv</code></p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>log_info</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, log intermediate progress information using the default logger. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.float64">float64</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.float64: Predicted difficulty of the MSA.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>pypythia/prediction.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">predict_difficulty</span><span class="p">(</span>
    <span class="n">msa_file</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">,</span>
    <span class="n">model_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="n">DEFAULT_MODEL_FILE</span><span class="p">,</span>
    <span class="n">raxmlng</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="n">DEFAULT_RAXMLNG_EXE</span><span class="p">,</span>
    <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">file_format</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">FileFormat</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">DataType</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">deduplicate</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">remove_full_gaps</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">result_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">store_results</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">log_info</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Predict the difficulty of an MSA using the PyPythia difficulty predictor.</span>

<span class="sd">    Per default, the MSA is deduplicated and full gap sequences are removed before the difficulty is predicted.</span>

<span class="sd">    Args:</span>
<span class="sd">        msa_file (pathlib.Path): Path to the MSA file. Note that the MSA file must be in either FASTA or PHYLIP format.</span>
<span class="sd">        model_file (pathlib.Path, optional): Path to the trained difficulty predictor model.</span>
<span class="sd">            Defaults to the latest model shipped with PyPythia.</span>
<span class="sd">        raxmlng (pathlib.Path, optional): Path to the RAxML-NG executable.</span>
<span class="sd">            If not set, uses the RAxML-NG binary found in the PATH environment variable.</span>
<span class="sd">        threads (int, optional): Number of threads to use for parallel parsimony tree inference. If not set, uses the</span>
<span class="sd">            RAxML-NG auto parallelization scheme.</span>
<span class="sd">        seed (int, optional): Random seed to use for the parsimony tree inference. Defaults to 0.</span>
<span class="sd">        file_format (FileFormat, optional): File format of the MSA file. Defaults to None. In this case, the file format</span>
<span class="sd">            is inferred based on the file content. See `pypythia.msa.parse_msa` for information on when this is required.</span>
<span class="sd">        data_type (DataType, optional): Data type of the MSA sequences. Defaults to None. In this case, the data type</span>
<span class="sd">            is inferred based on the file content. See `pypythia.msa.parse_msa` for information on when this is required.</span>
<span class="sd">        deduplicate (bool, optional): If True, remove duplicate sequences from the MSA. Defaults to True.</span>
<span class="sd">        remove_full_gaps (bool, optional): If True, remove full gap sequences from the MSA. Defaults to True.</span>
<span class="sd">        result_prefix (pathlib.Path, optional): Prefix for the result files. Defaults to None. In this case, the prefix</span>
<span class="sd">            is set to the MSA file name.</span>
<span class="sd">        store_results (bool, optional): If True, store intermediate results as file. Defaults to True.</span>
<span class="sd">            In this case, the following files are stored:</span>
<span class="sd">            - The reduced MSA in PHYLIP format (if duplicates or full gap sequences were removed) in `{result_prefix}.reduced.phy`</span>
<span class="sd">            - The inferred parsimony trees in Newick format in `{result_prefix}.pythia.trees`</span>
<span class="sd">            - The shapley values as waterfall plot in `{result_prefix}.shap.pdf`</span>
<span class="sd">            - The features and predicted difficulty as CSV file in `{result_prefix}.pythia.csv`</span>

<span class="sd">        log_info (bool, optional): If True, log intermediate progress information using the default logger. Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.float64: Predicted difficulty of the MSA.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">msa_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="k">raise</span> <span class="n">PyPythiaException</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The given MSA </span><span class="si">{</span><span class="n">msa_file</span><span class="si">}</span><span class="s2"> file does not exist.&quot;</span><span class="p">)</span>

    <span class="n">result_prefix</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">result_prefix</span><span class="p">)</span> <span class="k">if</span> <span class="n">result_prefix</span> <span class="k">else</span> <span class="n">msa_file</span>

    <span class="n">pars_trees_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">result_prefix</span><span class="si">}</span><span class="s2">.pythia.trees&quot;</span><span class="p">)</span>
    <span class="n">shap_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">result_prefix</span><span class="si">}</span><span class="s2">.shap.pdf&quot;</span><span class="p">)</span>
    <span class="n">results_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">result_prefix</span><span class="si">}</span><span class="s2">.pythia.csv&quot;</span><span class="p">)</span>

    <span class="c1"># We definitely need to store the reduced MSA somewhere for RAxML-NG</span>
    <span class="k">if</span> <span class="n">store_results</span><span class="p">:</span>
        <span class="c1"># If the user wants to keep the results, use the result_prefix</span>
        <span class="n">reduced_msa_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">result_prefix</span><span class="si">}</span><span class="s2">.reduced.phy&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Else, use a temporary file</span>
        <span class="n">reduced_msa_file</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span>
            <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.phy&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>
        <span class="p">)</span>

    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
        <span class="n">message</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Starting prediction for MSA </span><span class="si">{</span><span class="n">msa_file</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Init RAxML-NG</span>
    <span class="k">if</span> <span class="n">raxmlng</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">PyPythiaException</span><span class="p">(</span>
            <span class="s2">&quot;Path to the RAxML-NG executable is required if &#39;raxml-ng&#39; is not in $PATH.&quot;</span>
        <span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">raxmlng</span> <span class="o">=</span> <span class="n">RAxMLNG</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="s2">&quot;exe_path&quot;</span><span class="p">:</span> <span class="n">raxmlng</span><span class="p">}</span> <span class="k">if</span> <span class="n">raxmlng</span> <span class="k">else</span> <span class="p">{})</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">PyPythiaException</span><span class="p">(</span><span class="s2">&quot;Initializing RAxML-NG failed.&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">e</span>

    <span class="c1"># Init the prediction model</span>
    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span><span class="n">message</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Loading predictor </span><span class="si">{</span><span class="n">model_file</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">predictor</span> <span class="o">=</span> <span class="n">DifficultyPredictor</span><span class="p">(</span><span class="n">model_file</span><span class="o">=</span><span class="n">model_file</span><span class="p">)</span>

    <span class="c1"># Load the MSA</span>
    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span><span class="n">message</span><span class="o">=</span><span class="s2">&quot;Loading MSA&quot;</span><span class="p">)</span>

    <span class="n">msa</span> <span class="o">=</span> <span class="n">parse_msa</span><span class="p">(</span><span class="n">msa_file</span><span class="p">,</span> <span class="n">file_format</span><span class="o">=</span><span class="n">file_format</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>

    <span class="c1"># Deduplicate the MSA if necessary</span>
    <span class="n">reduced_msa</span> <span class="o">=</span> <span class="n">_handle_duplicates</span><span class="p">(</span><span class="n">msa</span><span class="p">,</span> <span class="n">deduplicate</span><span class="p">)</span>

    <span class="c1"># Remove full gap sequences if necessary</span>
    <span class="n">reduced_msa</span> <span class="o">=</span> <span class="n">_handle_full_gap_sequences</span><span class="p">(</span><span class="n">reduced_msa</span><span class="p">,</span> <span class="n">remove_full_gaps</span><span class="p">)</span>

    <span class="c1"># Check if the reduced MSA is different from the original MSA</span>
    <span class="n">is_reduced</span> <span class="o">=</span> <span class="n">msa</span> <span class="o">!=</span> <span class="n">reduced_msa</span>
    <span class="k">if</span> <span class="n">is_reduced</span><span class="p">:</span>
        <span class="c1"># If the reduced MSA is different from the original MSA, proceed with the reduced MSA</span>
        <span class="n">msa</span> <span class="o">=</span> <span class="n">reduced_msa</span>

        <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
            <span class="s2">&quot;The input MSA contained duplicate sequences and/or sequences containing only gaps. &quot;</span>
            <span class="s2">&quot;WARNING: This predicted difficulty is only applicable to the reduced MSA (duplicate sequences removed). &quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Save the reduced MSA</span>
        <span class="n">msa_file</span> <span class="o">=</span> <span class="n">reduced_msa_file</span>
        <span class="n">msa</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">msa_file</span><span class="p">)</span>

        <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Saving a reduced alignment as </span><span class="si">{</span><span class="n">reduced_msa_file</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="sa">f</span><span class="s2">&quot;We recommend to only use the reduced alignment </span><span class="si">{</span><span class="n">reduced_msa_file</span><span class="si">}</span><span class="s2"> for your subsequent analyses.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Compute the MSA Features</span>
    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Starting to compute MSA features for MSA </span><span class="si">{</span><span class="n">msa_file</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span>
        <span class="s2">&quot;Number of threads not specified, using RAxML-NG autoconfig.&quot;</span>
        <span class="k">if</span> <span class="n">threads</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;Using </span><span class="si">{</span><span class="n">threads</span><span class="si">}</span><span class="s2"> threads for parallel parsimony tree computation.&quot;</span>
    <span class="p">)</span>

    <span class="c1"># If the MSA/reduced MSA contains less than 4 sequences, RAxML-NG will fail as there is only a single possible</span>
    <span class="c1"># tree topology for this MSA. In this case, any phylogenetic inference is meaningless and we raise a</span>
    <span class="c1"># PyPythia exception to inform the user.</span>
    <span class="k">if</span> <span class="n">msa</span><span class="o">.</span><span class="n">n_taxa</span> <span class="o">&lt;</span> <span class="mi">4</span><span class="p">:</span>
        <span class="n">error_msg</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;The MSA contains less than 4 sequences. &quot;</span>
            <span class="s2">&quot;Phylogenetic inference is not meaningful for such small MSAs as there exists only a single possible tree topology. &quot;</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">is_reduced</span><span class="p">:</span>
            <span class="n">error_msg</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="s2">&quot;Note that during preprocessing, Pythia reduced the input MSA by removing duplicate sequences and/or &quot;</span>
                <span class="s2">&quot;sequences containing only gaps leading to an MSA with less than 4 sequences. &quot;</span>
                <span class="s2">&quot;You can rerun the prediction and disable deduplication and gap removal to use the original MSA. &quot;</span>
            <span class="p">)</span>
        <span class="k">raise</span> <span class="n">PyPythiaException</span><span class="p">(</span><span class="n">error_msg</span><span class="p">)</span>

    <span class="n">msa_features</span> <span class="o">=</span> <span class="n">collect_features</span><span class="p">(</span>
        <span class="n">msa</span><span class="o">=</span><span class="n">msa</span><span class="p">,</span>
        <span class="n">msa_file</span><span class="o">=</span><span class="n">msa_file</span><span class="p">,</span>
        <span class="n">raxmlng</span><span class="o">=</span><span class="n">raxmlng</span><span class="p">,</span>
        <span class="n">pars_trees_file</span><span class="o">=</span><span class="n">pars_trees_file</span> <span class="k">if</span> <span class="n">store_results</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">log_info</span><span class="o">=</span><span class="n">log_info</span><span class="p">,</span>
        <span class="n">threads</span><span class="o">=</span><span class="n">threads</span><span class="p">,</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Predict the difficulty</span>
    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span><span class="s2">&quot;Predicting the difficulty&quot;</span><span class="p">)</span>
    <span class="n">difficulty</span> <span class="o">=</span> <span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">msa_features</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">store_results</span><span class="p">:</span>
        <span class="c1"># Plot shapley values</span>
        <span class="c1"># this only makes sense if store_results=True, otherwise the figure would be lost</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">predictor</span><span class="o">.</span><span class="n">plot_shapley_values</span><span class="p">(</span><span class="n">msa_features</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fname</span><span class="o">=</span><span class="n">shap_file</span><span class="p">)</span>

    <span class="n">log_info</span> <span class="ow">and</span> <span class="n">log_runtime_information</span><span class="p">(</span><span class="s2">&quot;Done&quot;</span><span class="p">)</span>

    <span class="c1"># Log the feature values</span>
    <span class="k">if</span> <span class="n">log_info</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="o">*</span> <span class="mi">20</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;FEATURES: &quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">feat</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">msa_features</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">store_results</span><span class="p">:</span>
        <span class="c1"># Write the features + difficulty</span>
        <span class="n">msa_features</span><span class="p">[</span><span class="s2">&quot;difficulty&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">difficulty</span>
        <span class="n">msa_features</span><span class="p">[</span><span class="s2">&quot;msa_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">msa_file</span><span class="p">)</span>

        <span class="n">msa_features</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">results_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">log_info</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Results: </span><span class="si">{</span><span class="n">results_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">is_reduced</span> <span class="ow">and</span> <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reduced MSA: </span><span class="si">{</span><span class="n">reduced_msa_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Inferred parsimony trees: </span><span class="si">{</span><span class="n">pars_trees_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SHAP waterfall plot: </span><span class="si">{</span><span class="n">shap_file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;WARNING: When using shap plots, make sure you understand what shapley values are and how you can interpret&quot;</span>
            <span class="s2">&quot; this plot. For details refer to the wiki: https://github.com/tschuelia/PyPythia/wiki/Usage#shapley-values&quot;</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">difficulty</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>